scripts/qa_crowbarsetup.sh

```diff
@@ -5038,10 +5039,140 @@ function ping_fips
     done
 }
 
+function get_stack_fips
+{
+    local stack=$1
+    local fip_file=$2
+    local all_fips=$(mktemp)
+    local fip_uuids=$(openstack stack resource list $stack -f value -c physical_resource_id --filter type=OS::Neutron::FloatingIP)
+    openstack floating ip list -f value > $all_fips
+
+    for uuid in $fip_uuids
+        do
+        echo $(awk "\$1 ~ /${uuid}/ { print \$2 }" $all_fips) >> $fip_file
+        done
+
+    rm $all_fips
+}
+
+
+function ping_stack_fips
+{
+    . .openrc
+
+    local stack=$1
+    local fips=$(mktemp)
+
+    get_stack_fips $stack $fips
+
+    for fip in $(cat $fips)
+        do
+        ping -c 1 -w 60 $fip || complain 120 "cannot reach stack $stack floating IP $fip."
+        done
+
+    rm $fips
+}
+
+# Disable all but 3 compute nodes (these are the ones we will deploy the
+# pre-upgrade Heat stack on)
+function oncontroller_reduce_to_3_computes
+{
+    . .openrc
+    echo "Reducing cloud to 3 active compute nodes..."
+    local enable_computes=$(nova service-list | grep nova-compute | awk '{print $2}' | head -n 3)
```

Comment:
Why do we need to explicitly enable first three? All of them should be already enabled by default, right...

Reply:
"should". This ensures they are...

Reply:
True. But why shouldn't they be already, after deployment? Actually if they are not, it could mean deployment were wrong and we should not try to fix it but rather let it fail...

Reply:
Deployment should leave them enabled, yes. They might not be enabled if this code was run before and interrupted (it is designed to be run multiple times as needed) or if somebody enabled/disabled services manually.

Reply:
Well, OK. I find it unnecessary but let it be.